# Applied Data Science Capstone Assignment 2 : Segmenting and Clustering Neighborhoods in Toronto - PART 1

<b>For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment.
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:<br>
    --- Table on the Wikipedia Page linked in the Point 2 ---

<b>Additional Link:</b><br>
Link for scrapping a table from a website: https://towardsdatascience.com/web-scraping-html-tables-with-python-c9baba21059

<h4>Code Starts Here:</h4>

In [20]:
# Import necessary libraries

import requests
import lxml.html as lh
import pandas as pd

In [21]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [22]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This means that there are 3 columns per row

In [23]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postal Code
"
2:"Borough
"
3:"Neighborhood
"


<b>Creating Pandas DataFrame</b><p>
    Each header is appended to a tuple along with an empty list.

In [24]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [25]:
# Check the length of each column. Ideally, they should all be the same
[len(C) for (title,C) in col]

[181, 181, 181]

This shows that each of the 3 columns has exactly 181 rows

<b>Creating the pandas data frame

In [26]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

3. To create the above dataframe:

<ul>
    <li>The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood</li>
    <li>Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.</li>
    <li>More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.</li>
<li>If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.</li>
<li>Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.</li>
<li>In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.</li>
</ul>


4. Submit a link to your Notebook on your Github repository. (10 marks)

<b>Note:</b> There are different website scraping libraries and packages in Python. For scraping the above table, you can simply use pandas to read the table into a pandas dataframe.

Another way, which would help to learn for more complicated cases of web scraping is using the BeautifulSoup package. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [27]:
# Access the top 5 rows of the data frame 
df.head()

Postal Code\n           Borough\n               Neighborhood\n
0         M1A\n      Not assigned\n               Not assigned\n
1         M2A\n      Not assigned\n               Not assigned\n
2         M3A\n        North York\n                  Parkwoods\n
3         M4A\n        North York\n           Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

In [28]:
#Rearranging and renaming the columns

df.columns = ['Borough', 'Neighbourhood','Postcode']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df.head()

Postcode Borough       Neighbourhood
0               Not assigned\n   M1A\n      Not assigned\n
1               Not assigned\n   M2A\n      Not assigned\n
2                  Parkwoods\n   M3A\n        North York\n
3           Victoria Village\n   M4A\n        North York\n
4  Regent Park, Harbourfront\n   M5A\n  Downtown Toronto\n

In [29]:
#Cleaning the messy string in the Borough column

df = df.replace('\n',' ', regex=True)
df.head()

Postcode Borough      Neighbourhood
0               Not assigned     M1A       Not assigned 
1               Not assigned     M2A       Not assigned 
2                  Parkwoods     M3A         North York 
3           Victoria Village     M4A         North York 
4  Regent Park, Harbourfront     M5A   Downtown Toronto

In [30]:
#Dropping all cells with a borough that is Not assigned

df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode Borough      Neighbourhood
0                                 Not assigned     M1A       Not assigned 
1                                 Not assigned     M2A       Not assigned 
2                                    Parkwoods     M3A         North York 
3                             Victoria Village     M4A         North York 
4                    Regent Park, Harbourfront     M5A   Downtown Toronto 
5             Lawrence Manor, Lawrence Heights     M6A         North York 
6  Queen's Park, Ontario Provincial Government     M7A   Downtown Toronto 
7                                 Not assigned     M8A       Not assigned 
8      Islington Avenue, Humber Valley Village     M9A          Etobicoke 
9                               Malvern, Rouge     M1B        Scarborough

In [31]:
#Combining Neighbourhoods based on similar Postcode and Borough

df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head(10)

Postcode Borough  \
0                                                              
1                                         Agincourt     M1S    
2                            Alderwood, Long Branch     M8W    
3   Bathurst Manor, Wilson Heights, Downsview North     M3H    
4                                   Bayview Village     M2K    
5                 Bedford Park, Lawrence Manor East     M5M    
6                                       Berczy Park     M5E    
7                       Birch Cliff, Cliffside West     M1N    
8      Brockton, Parkdale Village, Exhibition Place     M6K    
9  Business reply mail Processing Centre, South C...    M7Y    

            Neighbourhood  
0  Canadian postal codes   
1            Scarborough   
2              Etobicoke   
3             North York   
4             North York   
5             North York   
6       Downtown Toronto   
7            Scarborough   
8           West Toronto   
9           East Toronto

In [32]:
#Removing any space in the start of the string

df['Neighbourhood'] = df['Neighbourhood'].str.strip()

In [33]:
#Assigning Borough values to the Neignbourhood where value is "Not assigned"

df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [34]:
# Check if the Neighbourhood for Queen's Park changed 

df[df['Borough'] == 'Queen\'s Park']

Empty DataFrame
Columns: [Postcode, Borough, Neighbourhood]
Index: []

In [35]:
# Check the shape of the data frame
df.shape

(181, 3)

In [36]:
#Save this file to a csv

df.to_csv(r'df_canada.csv')